assets/product-info.js

```diff
@@ -68,37 +69,30 @@ if (!customElements.get('product-info')) {
         });
       }
 
-      handleOptionValueChange({ data: { event, target, variant } }) {
+      handleOptionValueChange({ data: { event, target, selectedOptionValues } }) {
         if (!this.contains(event.target)) return;
 
-        const targetUrl = target.dataset.productUrl || this.dataset.url;
+        const baseUrl = target.dataset.productUrl || this.inflightRequestProductUrl || this.dataset.url;
+        this.inflightRequestProductUrl = baseUrl;
 
         const productForm = this.productForm;
         productForm?.toggleSubmitButton(true);
         productForm?.handleErrorMessage();
 
-        let callback = () => {};
-        let productUrl = this.getProductInfoUrl(targetUrl, variant?.id);
-        if (this.dataset.url !== targetUrl) {
-          callback = this.handleSwapProduct(targetUrl, variant);
-          productUrl = this.getProductInfoUrl(targetUrl, variant?.id, true);
-        } else if (!variant) {
-          this.setUnavailable();
-          callback = (html) => {
-            this.pickupAvailability?.update(variant);
-            this.updateOptionValues(html);
-          };
-        } else {
-          this.updateVariantInputs(variant.id);
-          callback = this.handleUpdateProductInfo(targetUrl, variant);
-        }
-
-        this.renderProductInfo(productUrl, target.id, callback);
+        const shouldSwapProduct = this.dataset.url !== baseUrl;
+        this.renderProductInfo({
+          productUrl: this.getProductInfoUrl(baseUrl, selectedOptionValues, shouldSwapProduct),
+          targetId: target.id,
+          callback: shouldSwapProduct ? this.handleSwapProduct(baseUrl) : this.handleUpdateProductInfo(baseUrl),
+        });
```

Comment:
This logic was refactored because we don't want to / can't access the `variant` entity until the section rendering request resolves.

Primary changes are:
- [Unavailable state](https://github.com/Shopify/dawn/pull/3495/files#diff-1ab5b2ed05e2f2ba4adfb4b617297e5d95e49a89d944f6e3ed33ad39d44f69ddL86-L90) is now deferred and [set in the request callback](https://github.com/Shopify/dawn/pull/3495/files#diff-1ab5b2ed05e2f2ba4adfb4b617297e5d95e49a89d944f6e3ed33ad39d44f69ddR166-R169)
- [Variant inputs](https://github.com/Shopify/dawn/pull/3495/files#diff-1ab5b2ed05e2f2ba4adfb4b617297e5d95e49a89d944f6e3ed33ad39d44f69ddL92) on the product form are now set [in the request callback](https://github.com/Shopify/dawn/pull/3495/files#diff-1ab5b2ed05e2f2ba4adfb4b617297e5d95e49a89d944f6e3ed33ad39d44f69ddR164)